safehttp/form.go

```diff
@@ -0,0 +1,234 @@
+	}
+	vals, ok := f.values[paramName]
+	if !ok {
+		f.err = errors.New("no value found for key " + paramName)
+		return defaultValue
+	}
+	paramVal, err := strconv.ParseFloat(vals[0], 64)
+	if err != nil {
+		f.err = err
+		return defaultValue
+	}
+	return paramVal
+}
+
+// Bool TODO
+func (f *Form) Bool(paramName string, defaultValue bool) bool {
+	if f.err != nil {
+		return defaultValue
+	}
+	vals, ok := f.values[paramName]
+	if !ok {
+		f.err = errors.New("no value found for key " + paramName)
+		return defaultValue
+	}
+	if vals[0] != "true" {
+		if vals[0] != "false" {
+			f.err = errors.New("values of form parameter " + paramName + " not a boolean")
+		}
+		return false
+	}
+	return true
+}
+
+// Slice TODO
+func (f *Form) Slice(slice interface{}, paramName string) {
+	mapVals, ok := f.values[paramName]
+	switch values := slice.(type) {
+	case *[]string:
+		if f.err != nil {
+			*values = nil
+			return
+		}
+		if !ok {
+			f.err = errors.New("no value found for key " + paramName)
+			*values = nil
+		}
+		res := make([]string, 0, len(mapVals))
+		for _, x := range mapVals {
+			res = append(res, x)
+		}
+		*values = res
+	case *[]int:
+		if f.err != nil {
+			*values = nil
+			return
+		}
+		if !ok {
+			f.err = errors.New("no value found for key " + paramName)
+			*values = nil
+		}
+		res := make([]int, 0, len(mapVals))
+		for _, x := range mapVals {
+			x, err := strconv.Atoi(x)
+			if err != nil {
+				f.err = err
+				*values = nil
+				return
+			}
+			res = append(res, x)
+			*values = res
+		}
+	case *[]uint64:
+		if f.err != nil {
+			*values = nil
+			return
+		}
+		if !ok {
+			f.err = errors.New("no value found for key " + paramName)
+			*values = nil
+		}
+		res := make([]uint64, 0, len(mapVals))
+		for _, x := range mapVals {
+			x, err := strconv.ParseUint(x, 10, 0)
+			if err != nil {
+				f.err = err
+				*values = nil
+				return
+			}
+			res = append(res, x)
+		}
+		*values = res
+	case *[]float64:
+		if f.err != nil {
+			*values = nil
+			return
+		}
+		if !ok {
+			f.err = errors.New("no value found for key " + paramName)
+			*values = nil
+		}
+		res := make([]float64, 0, len(mapVals))
+		for _, x := range mapVals {
+			x, err := strconv.ParseFloat(x, 64)
+			if err != nil {
+				f.err = err
+				*values = nil
+				return
+			}
+			res = append(res, x)
+		}
+		*values = res
+	case *[]bool:
+		if f.err != nil {
+			*values = nil
+			return
+		}
+		if !ok {
+			f.err = errors.New("no value found for key " + paramName)
+			*values = nil
+		}
+		res := make([]bool, 0, len(mapVals))
+		for _, x := range mapVals {
+			if x != "true" {
+				if x != "false" {
+					f.err = errors.New(": values of form parameter " + paramName + " not a boolean")
+					*values = nil
+					return
+				}
+				res = append(res, false)
+				continue
+			}
+			res = append(res, true)
+		}
+		*values = res
+	default:
+		f.err = errors.New("slice type not supported")
+	}
+	return
+}
+
+// Error TODO
+func (f *Form) Error() error {
+	return f.err
+}
```

Comment:
I'm not sure we want Form to implement the `error` interface. Maybe call this `Err`?

Reply:
Resolved in [d511d47](url)